### DISCUSSION TITLE
What does "data-driven" mean?

### DISCUSSION THREAD

**Discussion by kartikynwa** [bevyengine/bevy#5941] [Q&A] [answered]
Asking this out of curiosity as someone who has no knowledge of game development.

I looked at the documentation which says this:

> Data Focused: Data-oriented architecture using the Entity Component System paradigm

And the homepage says:

> All engine and game logic uses Bevy ECS, a custom Entity Component System 

I still don't understand it. Maybe a comparison with a non-data-driven system will help.

Also I am not implying that the documentation should be improved or anything. Just asking out of curiosity as mentioned earlier. I hope that is okay.

**Comment by harudagondi** [comment, marked as answer]
This isn't related to Bevy, but this [resource][bracketlib] on making a roguelike using [`specs`] (an older ECS library) shows how making data-driven design makes your code easier to mod and edit, and decouples game logic from game data.

You should read [Section 5] of that book, especially the second part titled ["Data-Driven Design: Raw Files"][rawfiles]. It shows examples of data-driven design logic.

[bracketlib]: https://bfnightly.bracketproductions.com/rustbook/chapter_0.html
[`specs`]: https://github.com/amethyst/specs
[Section 5]: https://bfnightly.bracketproductions.com/rustbook/chapter_43.html
[rawfiles]: https://bfnightly.bracketproductions.com/rustbook/chapter_45.html

**Reply by kartikynwa** [reply]
Thanks a lot

**Comment by Nilirad** [comment]
Data-driven essentially means, to my understanding, that the behavior of a program is mostly controlled by data (like assets, scenes), instead of being hardcoded.

Bevy is not fully data driven, since e.g. systems can only be defined with Rust code. But for the most part it can be considered to be data driven.

See also bevyengine/bevy-website#81

**Reply by kartikynwa** [reply]
Thanks :)
